// langchain/dist/chat_models/base.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleChatModel = exports.BaseChatModel = void 0;
const index_js_1 = require("../schema/index.cjs");
const index_js_2 = require("../base_language/index.cjs");
const base_js_1 = require("../memory/base.cjs");
class BaseChatModel extends index_js_2.BaseLanguageModel {
    constructor(fields) {
        super(fields);
    }
    async generate(messages, stop) {
        const generations = [];
        const llmOutputs = [];
        const messageStrings = messages.map((messageList) => (0, base_js_1.getBufferString)(messageList));
        await this.callbackManager.handleLLMStart({ name: this._llmType() }, messageStrings, this.verbose);
        try {
            for (const message of messages) {
                const result = await this._generate(message, stop);
                if (result.llmOutput) {
                    llmOutputs.push(result.llmOutput);
                }
                generations.push(result.generations);
            }
        }
        catch (err) {
            await this.callbackManager.handleLLMError(err, this.verbose);
            throw err;
        }
        const output = {
            generations,
            llmOutput: llmOutputs.length
                ? this._combineLLMOutput?.(...llmOutputs)
                : undefined,
        };
        await this.callbackManager.handleLLMEnd(output, this.verbose);
        return output;
    }
    _modelType() {
        return "base_chat_model";
    }
    async generatePrompt(promptValues, stop) {
        const promptMessages = promptValues.map((promptValue) => promptValue.toChatMessages());
        return this.generate(promptMessages, stop);
    }
    async call(messages, stop) {
        const result = await this.generate([messages], stop);
        const generations = result.generations;
        return generations[0][0].message;
    }
    async callPrompt(promptValue, stop) {
        const promptMessages = promptValue.toChatMessages();
        return this.call(promptMessages, stop);
    }
}
exports.BaseChatModel = BaseChatModel;
class SimpleChatModel extends BaseChatModel {
    async _generate(messages, stop) {
        const text = await this._call(messages, stop);
        const message = new index_js_1.AIChatMessage(text);
        return {
            generations: [
                {
                    text: message.text,
                    message,
                },
            ],
        };
    }
}
exports.SimpleChatModel = SimpleChatModel;
