// langchain/dist/output_parsers/structured.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StructuredOutputParser = void 0;
/* eslint-disable no-instanceof/no-instanceof */
const zod_1 = require("zod");
const index_js_1 = require("../schema/index.cjs");
function printSchema(schema, depth = 0) {
    if (schema instanceof zod_1.z.ZodString &&
        schema._def.checks.some((check) => check.kind === "datetime")) {
        return "datetime";
    }
    if (schema instanceof zod_1.z.ZodString) {
        return "string";
    }
    if (schema instanceof zod_1.z.ZodNumber) {
        return "number";
    }
    if (schema instanceof zod_1.z.ZodBoolean) {
        return "boolean";
    }
    if (schema instanceof zod_1.z.ZodDate) {
        return "date";
    }
    if (schema instanceof zod_1.z.ZodNullable) {
        return `${printSchema(schema._def.innerType, depth)} // Nullable`;
    }
    if (schema instanceof zod_1.z.ZodTransformer) {
        return `${printSchema(schema._def.schema, depth)}`;
    }
    if (schema instanceof zod_1.z.ZodOptional) {
        return `${printSchema(schema._def.innerType, depth)} // Optional`;
    }
    if (schema instanceof zod_1.z.ZodArray) {
        return `${printSchema(schema._def.type, depth)}[]`;
    }
    if (schema instanceof zod_1.z.ZodObject) {
        const indent = "\t".repeat(depth);
        const indentIn = "\t".repeat(depth + 1);
        return `{${schema._def.description ? ` // ${schema._def.description}` : ""}
${Object.entries(schema.shape)
            .map(([key, value]) => `${indentIn}"${key}": ${printSchema(value, depth + 1)}${value._def.description
            ? ` // ${value._def.description}`
            : ""}`)
            .join("\n")}
${indent}}`;
    }
    throw new Error(`Unsupported type: ${schema._def.innerType.typeName}`);
}
class StructuredOutputParser extends index_js_1.BaseOutputParser {
    constructor(schema) {
        super();
        Object.defineProperty(this, "schema", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: schema
        });
    }
    static fromZodSchema(schema) {
        return new this(schema);
    }
    static fromNamesAndDescriptions(schemas) {
        const zodSchema = zod_1.z.object(Object.fromEntries(Object.entries(schemas).map(([name, description]) => [name, zod_1.z.string().describe(description)])));
        return new this(zodSchema);
    }
    getFormatInstructions() {
        return `The output should be a markdown code snippet formatted in the following schema:

\`\`\`json
${printSchema(this.schema)}
\`\`\`

Including the leading and trailing "\`\`\`json" and "\`\`\`"
`;
    }
    async parse(text) {
        try {
            const json = text.trim().split("```json")[1].split("```")[0].trim();
            return this.schema.parse(JSON.parse(json));
        }
        catch (e) {
            throw new index_js_1.OutputParserException(`Failed to parse. Text: "${text}". Error: ${e}`);
        }
    }
}
exports.StructuredOutputParser = StructuredOutputParser;
