// langchain/dist/memory/chat_memory.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseChatMemory = exports.ChatMessageHistory = void 0;
const index_js_1 = require("../schema/index.cjs");
const base_js_1 = require("./base.cjs");
class ChatMessageHistory extends index_js_1.BaseChatMessageHistory {
    constructor(messages) {
        super();
        Object.defineProperty(this, "messages", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        this.messages = messages ?? [];
    }
    addUserMessage(message) {
        this.messages.push(new index_js_1.HumanChatMessage(message));
    }
    addAIChatMessage(message) {
        this.messages.push(new index_js_1.AIChatMessage(message));
    }
}
exports.ChatMessageHistory = ChatMessageHistory;
class BaseChatMemory extends base_js_1.BaseMemory {
    constructor(fields) {
        super();
        Object.defineProperty(this, "chatHistory", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "returnMessages", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        Object.defineProperty(this, "inputKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "outputKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.chatHistory = fields?.chatHistory ?? new ChatMessageHistory();
        this.returnMessages = fields?.returnMessages ?? this.returnMessages;
        this.inputKey = fields?.inputKey ?? this.inputKey;
        this.outputKey = fields?.outputKey ?? this.outputKey;
    }
    async saveContext(inputValues, outputValues) {
        this.chatHistory.addUserMessage((0, base_js_1.getInputValue)(inputValues, this.inputKey));
        this.chatHistory.addAIChatMessage((0, base_js_1.getInputValue)(outputValues, this.outputKey));
    }
}
exports.BaseChatMemory = BaseChatMemory;
